### ISSUE TITLE
Active Record value casting can cause unexpected behavior with Time values

### ISSUE THREAD

**Issue by aprescott** [jackc/tod#63]
* Rails 5.2.2.1
* tod version 2.2.0

I work on an application which uses the tod gem. A test started failing today (April 2nd) because of some unexpected behavior when assigning a `Time` value.

## Summary

```ruby
>> today = Date.parse("2019-04-02")
#=> Tue, 02 Apr 2019
>> schedule.open_at = today.beginning_of_day
#=> Tue, 02 Apr 2019 00:00:00 EDT -04:00
>> schedule.close_at = today.end_of_day
=> Tue, 02 Apr 2019 23:59:59 EDT -04:00
>> [schedule.open_at.to_s, schedule.close_at.to_s]
#=> ["24:00:00", "23:59:59"]
```

It looks like the `"24:00:00"` value here happens only on the 2nd of the month, and on other days the result would be `["00:00:00", "23:59:59"]`:

```ruby
>> today = Date.parse("2019-04-03")
#=> Wed, 03 Apr 2019
>> schedule.open_at = today.beginning_of_day
#=> Wed, 03 Apr 2019 00:00:00 EDT -04:00
>> schedule.close_at = today.end_of_day
#=> Wed, 03 Apr 2019 23:59:59 EDT -04:00
>> [schedule.open_at.to_s, schedule.close_at.to_s]
#=> ["00:00:00", "23:59:59"]
```

## Further details

We have an AR model representing a schedule. It has 2 attributes backed by `time without time zone`:

```sql
-- structure.sql
open_at time without time zone NOT NULL,
close_at time without time zone NOT NULL,
```

Because of the way [AR casting works with `TimeOfDayType`](https://github.com/jackc/tod/blob/v2.2.0/lib/tod/time_of_day_type.rb#L3-L5), `TimeOfDay.load` is called when assigning attributes. In the case where the RHS value is a `Time`, this can lead to unexpected behavior because of [a `time.day == 2` conditional in `TimeOfDay.load`](https://github.com/jackc/tod/blob/v2.2.0/lib/tod/time_of_day.rb#L219):

```ruby
>> today = Date.parse("2019-04-02")
#=> Tue, 02 Apr 2019
>> schedule.open_at = today.beginning_of_day
#=> Tue, 02 Apr 2019 00:00:00 EDT -04:00
>> schedule.close_at = today.end_of_day
=> Tue, 02 Apr 2019 23:59:59 EDT -04:00
>> [schedule.open_at.to_s, schedule.close_at.to_s]
#=> ["24:00:00", "23:59:59"]
```

I think this is pretty unexpected, because the behavior is different for other days, like the 3rd of a month:

```ruby
>> today = Date.parse("2019-04-03")
#=> Wed, 03 Apr 2019
>> schedule.open_at = today.beginning_of_day
#=> Wed, 03 Apr 2019 00:00:00 EDT -04:00
>> schedule.close_at = today.end_of_day
#=> Wed, 03 Apr 2019 23:59:59 EDT -04:00
>> [schedule.open_at.to_s, schedule.close_at.to_s]
#=> ["00:00:00", "23:59:59"]
```

You can see that on April 2nd the result is `["24:00:00", "23:59:59"]`, but on April 3rd the result is `["00:00:00", "23:59:59"]`.

You can see this more directly by calling methods directly:

```ruby
>> Tod::TimeOfDay.load(Date.parse("2019-03-26").in_time_zone("US/Eastern").tap { |t| p t }).to_s
Tue, 26 Mar 2019 00:00:00 EDT -04:00
#=> "00:00:00"
>> Tod::TimeOfDay.load(Date.parse("2019-04-02").in_time_zone("US/Eastern").tap { |t| p t }).to_s
Tue, 02 Apr 2019 00:00:00 EDT -04:00
#=> "24:00:00"
```

I'm not entirely sure why there's a [`time.day == 2`](https://github.com/jackc/tod/blob/v2.2.0/lib/tod/time_of_day.rb#L219) check, but my guess is that there's an assumption somewhere that time values will mostly have `day == 1`, so that midnight on `day == 2` can represent 24:00.

**Comment by jackc** [comment]
Hmm... That appears to be an unfortunate interaction with code that tries to allow persisting 24:00:00 as a time of day. If I recall correctly, the check for day == 2 is because of trying to work around how Rails handles database `time` types. Rails loads DB `time`  into a Ruby `Time`. So `24:00:00` gets loaded into something like `2019-04-02 00:00:00`.

It's unclear to me what would be total solution to this.

But you may be able to work around the issue. It appears that you are assigning Ruby `Time` objects to a `TimeOfDay` attribute on your `Schedule` model. This is probably triggering ActiveRecord's type conversion. If you changed `schedule.open_at = today.beginning_of_day` to `schedule.open_at = today.beginning_of_day.to_time_of_day` I suspect it would work.

**Comment by aprescott** [comment]
In the Active Record context, would it make sense to call `to_time_of_day` on assigned values when appropriate (e.g., when it's a `Time`), so that it's handled automatically?

Perhaps in general there should be stricter handling of `Time` values when it comes to methods like `TimeOfDay.load`? I'm trying to think about other non-AR situations where this could come up, but I'm not familiar enough with the use cases.

**Comment by jackc** [comment]
IIRC the underlying complication is that when Rails reads a `time` column from the database it converts it to a Ruby time **before** it gets to `tod`. So instead of parsing a string `24:00:00` it gets a Time of `yyyy-mm-02 00:00:00`. So in order to handle `24:00:00` there is a special case for the second day of the month when a datetime is converted. 

The problem appears to be the same type conversion code path happens when reading a value from the database and when assigning to an attribute.

**Comment by jackc** [comment]
I think that this is resolved in v3.0.0. It uses ActiveRecord attributes instead of serializable.
